### ISSUE TITLE
Support pachctl commands to TLS-enabled pachd when port forwarding

### ISSUE THREAD

**Issue by jfhamlin** [pachyderm/pachyderm#4465]
**What is the goal / desired outcome?**

It would be great to be able to run `pachctl` commands against a pachyderm cluster with TLS enabled without having to make pachd directly addressable (i.e. reachable without kube port forwarding).

**If there is a way to accomplish this today via workaround, what does that require?**

As far as I can tell, there is no workaround. Either TLS has to be disabled, or `pachd` needs to be reachable without port forwarding.

**(Optional) What is your proposal for a feature to solve this?**

1. Support setting `ServerCAs` without setting `PachdAddress`. This should cause the client to assume a `Secured` address but fall back to port forwarding to connect to the daemon.
2. To allow certificate verification to succeed when connecting to localhost, support a new `ServerNameOverride` or `PachdServerNameOverride` context setting. This will be passed to [credentials.NewClientTLSFromCert ](https://github.com/pachyderm/pachyderm/blob/master/src/client/client.go#L559) to allow certificate verification to proceed despite a hostname mismatch.

**Environment?**:
Kubernetes version (use kubectl version): v1.14.7
Pachyderm CLI and pachd server version (use pachctl version): 1.9.10
Cloud provider (e.g. aws, azure, gke) or local deployment (e.g. minikube vs dockerized k8s): aws (eks)
OS (e.g. from /etc/os-release):
Others:

**Comment by msteffen** [comment]
Hi James, thanks for filing this! This is definitely something we can add to our roadmap, but as a short-term workaround, you might be able to run `pachctl port-forward` and set `PachdAddress` to `grpcs://localhost:30650` (IIRC, that's the port that `port-forward` forwards) to get more or less this behavior.

Let us know if you try it, and if it works!

**Comment by jfhamlin** [comment]
> Let us know if you try it, and if it works!

Yep, I tried this! Thinking about it again, I can probably get the port-forward workaround to work by adding a line to my `hosts` file mapping the hostname in the certificate to 127.0.1.1. I still think it'd be great for `pachd` to support this configuration, as it seems like it would be a common case for TLS deployments. I actually ended up hacking this in and running a patched pachctl locally, so if there's alignment on the proposal I can submit a PR sometime soon.

**Comment by msteffen** [comment]
Oh interesting. I struck out my last response because I didn't think of modifying the hosts file.

That said, If I understand correctly, Kubernetes port-forwarding [already uses TLS](https://stackoverflow.com/questions/50645059/is-kubectl-port-forward-encrypted). In particular, any port-forwarded traffic will be tunneled through a connection between the kubernetes client and the Kubernetes API server, which is authenticated/encrypted using the generated certificate stored in your local kubeconfig file.

Is my understanding incorrect? If not, what makes you interested in connecting over TLS if that connection will be tunnelled through an existing TLS connection?

**Comment by jfhamlin** [comment]
Right, I believe it looks something like this:
```
[pachctl]<2----[k8s client]<1===(tunnel)===1>[k8s server]----2>[pachd]
```
... where the tunnel (1) is encrypted (handled by k8s), and the tunneled connection (2) is encrypted IFF pachyderm was deployed with TLS enabled. In this case I've deployed pachyderm with TLS, so I don't think I have a choice — to `pachctl` it looks like it's talking directly to `pachd`, which is only accepting TLS connections.

**Comment by msteffen** [comment]
@jfhamlin We're discussing some of our options internally (we've had various ideas related to encrypting all Pachyderm traffic by default) but I'd like to clarify one point:

> In this case I've deployed pachyderm with TLS, so I don't think I have a choice

It should be possible (by redeploying pachyderm, or even editing the pachd deployment to remove the `pachd-tls-cert` volume, though I haven't tested this yet) to disable TLS in Pachyderm. **Would you consider that acceptable, or are you trying to ensure that traffic between the k8s APIServer and pachd is encrypted?**

I ask because the first part of the diagram (pachctl <-> k8s client) happens either within pachctl (default port-forwarding) or over localhost (`pachctl port-forward`), so our sense is that encrypting that leg isn't so important. Encrypting intra-cluster traffic (api server <-> pachd) seems more important, but I think there are several ways of doing that, so we want to be sure that this one makes the most sense (Istio, for example, is a network fabric that I believe encrypts all intra-cluster communication).

In general, many users move away from port-forwarding for a variety for reasons, including that it's significantly lower throughput than connecting directly, which limits the rate at which data can be ingressed. Thus the situation you're describing struck us as an unusual combination of a fairly advanced requirement (needing to encrypt intra-cluster traffic) and a more relaxed, exporatory requirement (wanting to connect over port-forward). So we wanted to make sure we understood the problem you were trying to solve before jumping to solutions.

**Comment by jfhamlin** [comment]
> It should be possible (by redeploying pachyderm, or even editing the pachd deployment to remove the pachd-tls-cert volume, though I haven't tested this yet) to disable TLS in Pachyderm.

It's true I could choose to disable pachyderm TLS, but I'd like to keep TLS enabled :). The traffic I care about encrypting is indeed intra-cluster. I'd be happy to discuss my use case in more detail in the Pachyderm slack workspace, but generally speaking, I'm trying to test a cloud pachyderm deployment with TLS and minimal additional dependencies. It seems just a hair away from working without custom network configuration.

**Comment by chainlink** [comment]
I think you can port-forward on the peer port and it will work correctly
